File: windows/VoiceKey.Core/VocabularyEditor.cs
namespace VoiceKey.Core;

/// <summary>One "when I say X, type Y" row. Either half may be blank while editing.</summary>
public sealed record Replacement(string Spoken = "", string Written = "")
{
    internal bool IsComplete =>
        !string.IsNullOrWhiteSpace(Spoken) && !string.IsNullOrWhiteSpace(Written);
}

/// <summary>
/// The dictionary as an editable list of rows — what the Dictionary pane shows
/// and edits, and the only place that knows how rows become a <see cref="VocabularyDictionary"/>.
///
/// Rows exist because <see cref="VocabularyDictionary.Replacements"/> is an unordered map: a
/// table needs stable positions, and a row being edited is allowed to be blank or
/// duplicated for as long as the user is typing. Nothing invalid reaches disk —
/// <see cref="Dictionary"/> is the filter.
/// </summary>
public sealed class VocabularyEditor : IEquatable<VocabularyEditor>
{
    public IReadOnlyList<Replacement> Replacements { get; }
    public IReadOnlyList<string> Terms { get; }

    public VocabularyEditor() : this(new VocabularyDictionary()) { }

    public VocabularyEditor(VocabularyDictionary dictionary)
    {
        Replacements = dictionary.Replacements
            .OrderBy(entry => entry.Key, StringComparer.OrdinalIgnoreCase)
            .Select(entry => new Replacement(entry.Key, entry.Value))
            .ToArray();
        Terms = dictionary.Terms;
    }

    private VocabularyEditor(IReadOnlyList<Replacement> replacements, IReadOnlyList<string> terms)
    {
        Replacements = replacements;
        Terms = terms;
    }

    /// <summary>
    /// The saveable dictionary: blank halves and rows shadowed by an earlier row
    /// with the same spoken phrase are dropped, so what the user sees first wins.
    /// </summary>
    public VocabularyDictionary Dictionary
    {
        get
        {
            var ignored = IgnoredReplacementRows;
            var rules = Replacements
                .Select((row, index) => (row, index))
                .Where(entry => entry.row.IsComplete && !ignored.Contains(entry.index))
                .ToDictionary(entry => entry.row.Spoken.Trim(), entry => entry.row.Written.Trim());
            return new VocabularyDictionary(
                Terms.Select(term => term.Trim()).Where(term => term.Length > 0).ToArray(),
                rules);
        }
    }

    /// <summary>
    /// Rows that will not take effect because an earlier row already claims the
    /// same spoken phrase — matching is case-insensitive, so replacing is too.
    /// </summary>
    public IReadOnlySet<int> IgnoredReplacementRows
    {
        get
        {
            var claimed = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
            var ignored = new HashSet<int>();
            for (var index = 0; index < Replacements.Count; index++)
            {
                var spoken = Replacements[index].Spoken.Trim();
                if (spoken.Length == 0) continue;
                if (!claimed.Add(spoken)) ignored.Add(index);
            }
            return ignored;
        }
    }

    // MARK: - Editing

    public VocabularyEditor AddingReplacement() =>
        new([.. Replacements, new Replacement()], Terms);

    public VocabularyEditor SettingSpoken(string text, int index) =>
        UpdatingReplacement(index, row => row with { Spoken = text });

    public VocabularyEditor SettingWritten(string text, int index) =>
        UpdatingReplacement(index, row => row with { Written = text });

    public VocabularyEditor RemovingReplacements(IEnumerable<int> indexes) =>
        new(Removing(Replacements, indexes), Terms);

    public VocabularyEditor AddingTerm() => new(Replacements, [.. Terms, ""]);

    public VocabularyEditor SettingTerm(string text, int index)
    {
        if (index < 0 || index >= Terms.Count) return this;
        var updated = Terms.ToArray();
        updated[index] = text;
        return new VocabularyEditor(Replacements, updated);
    }

    public VocabularyEditor RemovingTerms(IEnumerable<int> indexes) =>
        new(Replacements, Removing(Terms, indexes));

    private VocabularyEditor UpdatingReplacement(int index, Func<Replacement, Replacement> edit)
    {
        if (index < 0 || index >= Replacements.Count) return this;
        var updated = Replacements.ToArray();
        updated[index] = edit(updated[index]);
        return new VocabularyEditor(updated, Terms);
    }

    private static T[] Removing<T>(IReadOnlyList<T> items, IEnumerable<int> indexes)
    {
        var dropped = indexes.ToHashSet();
        return items.Where((_, index) => !dropped.Contains(index)).ToArray();
    }

    // MARK: - Equality

    public bool Equals(VocabularyEditor? other) =>
        other is not null
        && Replacements.SequenceEqual(other.Replacements)
        && Terms.SequenceEqual(other.Terms);

    public override bool Equals(object? obj) => Equals(obj as VocabularyEditor);

    public override int GetHashCode() => HashCode.Combine(Replacements.Count, Terms.Count);
}

File: macos/Sources/VoiceKeyCore/VocabularyEditor.swift
import Foundation

/// The dictionary as an editable list of rows — what the Dictionary window shows
/// and edits, and the only place that knows how rows become a `VocabularyDictionary`.
///
/// Rows exist because `VocabularyDictionary.replacements` is an unordered map: a
/// table needs stable positions, and a row being edited is allowed to be blank or
/// duplicated for as long as the user is typing. Nothing invalid reaches disk —
/// `dictionary` is the filter.
public struct VocabularyEditor: Equatable {

    /// One "when I say X, type Y" row. Either half may be blank while editing.
    public struct Replacement: Equatable {
        public var spoken: String
        public var written: String

        public init(spoken: String = "", written: String = "") {
            self.spoken = spoken
            self.written = written
        }

        var isComplete: Bool { !spoken.trimmed.isEmpty && !written.trimmed.isEmpty }
    }

    public let replacements: [Replacement]
    public let terms: [String]

    public init(dictionary: VocabularyDictionary = VocabularyDictionary()) {
        self.replacements = dictionary.replacements
            .sorted { $0.key.localizedStandardCompare($1.key) == .orderedAscending }
            .map { Replacement(spoken: $0.key, written: $0.value) }
        self.terms = dictionary.terms
    }

    private init(replacements: [Replacement], terms: [String]) {
        self.replacements = replacements
        self.terms = terms
    }

    /// The saveable dictionary: blank halves and rows shadowed by an earlier row
    /// with the same spoken phrase are dropped, so what the user sees first wins.
    public var dictionary: VocabularyDictionary {
        let ignored = ignoredReplacementRows
        let rules = replacements.enumerated()
            .filter { $0.element.isComplete && !ignored.contains($0.offset) }
            .reduce(into: [String: String]()) { result, entry in
                result[entry.element.spoken.trimmed] = entry.element.written.trimmed
            }
        return VocabularyDictionary(terms: terms.map(\.trimmed).filter { !$0.isEmpty },
                                    replacements: rules)
    }

    /// Rows that will not take effect because an earlier row already claims the
    /// same spoken phrase — matching is case-insensitive, so replacing is too.
    public var ignoredReplacementRows: IndexSet {
        var claimed = Set<String>()
        return replacements.enumerated().reduce(into: IndexSet()) { ignored, entry in
            let spoken = entry.element.spoken.trimmed.lowercased()
            guard !spoken.isEmpty else { return }
            if !claimed.insert(spoken).inserted { ignored.insert(entry.offset) }
        }
    }

    // MARK: - Editing

    public func addingReplacement() -> VocabularyEditor {
        VocabularyEditor(replacements: replacements + [Replacement()], terms: terms)
    }

    public func settingSpoken(_ text: String, at index: Int) -> VocabularyEditor {
        updatingReplacement(at: index) { $0.spoken = text }
    }

    public func settingWritten(_ text: String, at index: Int) -> VocabularyEditor {
        updatingReplacement(at: index) { $0.written = text }
    }

    public func removingReplacements(at indexes: IndexSet) -> VocabularyEditor {
        VocabularyEditor(replacements: replacements.removing(indexes), terms: terms)
    }

    public func addingTerm() -> VocabularyEditor {
        VocabularyEditor(replacements: replacements, terms: terms + [""])
    }

    public func settingTerm(_ text: String, at index: Int) -> VocabularyEditor {
        guard terms.indices.contains(index) else { return self }
        var updated = terms
        updated[index] = text
        return VocabularyEditor(replacements: replacements, terms: updated)
    }

    public func removingTerms(at indexes: IndexSet) -> VocabularyEditor {
        VocabularyEditor(replacements: replacements, terms: terms.removing(indexes))
    }

    private func updatingReplacement(
        at index: Int,
        _ edit: (inout Replacement) -> Void
    ) -> VocabularyEditor {
        guard replacements.indices.contains(index) else { return self }
        var updated = replacements
        edit(&updated[index])
        return VocabularyEditor(replacements: updated, terms: terms)
    }
}

private extension Array {
    func removing(_ indexes: IndexSet) -> [Element] {
        enumerated().filter { !indexes.contains($0.offset) }.map(\.element)
    }
}

private extension String {
    var trimmed: String { trimmingCharacters(in: .whitespacesAndNewlines) }
}
